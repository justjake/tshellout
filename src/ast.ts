import { Buffer } from 'node:buffer'
import { PathLike } from 'node:fs'

export type Command =
  | { type: 'spawn'; executable: string; args: string[] }
  | { type: 'pipe'; lhs: Command; rhs: Command }
  | { type: 'and'; lhs: Command; rhs: Command }
  | { type: 'or'; lhs: Command; rhs: Command }
  | { type: 'redirectStdout'; cmd: Command; path: PathLike }
  | { type: 'redirectStderr'; cmd: Command; path: PathLike }
  | { type: 'redirectStdin'; cmd: Command; path: PathLike }
  | { type: 'writeStdin'; cmd: Command; data: Buffer }
