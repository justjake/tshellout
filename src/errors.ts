export class TShellOutError extends Error {}

export class NonZeroExitCode extends TShellOutError {
  code: number

  constructor(code: number) {
    super(`Process exited with non-zero error code: ${code}`)

    this.code = code
  }
}

export class ProcessTerminated extends TShellOutError {
  constructor(signal: NodeJS.Signals) {
    super(`Process was terminated by signal: ${signal}`)
  }
}
