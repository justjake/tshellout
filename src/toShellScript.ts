import { Command } from './ast'

function unreachable(x: never): never {
  throw new Error(`Unreachable: ${JSON.stringify(x)}`)
}

export function toShellScript(command: Command): string {
  switch (command.type) {
    case 'spawn':
      return [command.executable, ...command.args.map(quote)].join(' ')
    case 'pipe':
      return `${toShellScript(command.lhs)} | ${toShellScript(command.rhs)}`
    case 'and':
      return `(${toShellScript(command.lhs)}) && ${toShellScript(command.rhs)}`
    case 'or':
      return `(${toShellScript(command.lhs)}) || ${toShellScript(command.rhs)}`
    case 'redirectStdout':
      return `(${toShellScript(command.cmd)}) > ${quote(String(command.path))}`
    case 'redirectStderr':
      return `(${toShellScript(command.cmd)}) 2> ${quote(String(command.path))}`
    case 'redirectStdin':
      return `(${toShellScript(command.cmd)}) < ${quote(String(command.path))}`
    case 'writeStdin': {
      // Not sure how safe it is to pass binary data through quoting, etc.
      // Instead, pass as base64 decoded by the shell.
      const dataBase64 = command.data.toString('base64')
      return `${toShellScript(
        command.cmd
      )} <<< "$(base64 -d <<< ${dataBase64})"`
    }
    default:
      unreachable(command)
  }
}

/** Makes string `s` safe to interpolate as a shell argument. */
export function quote(s: string) {
  if (s === '') return `''`
  if (!/[^%+,-./:=@_0-9A-Za-z]/.test(s)) return s
  return `'` + s.replace(/'/g, `'"'`) + `'`
}
