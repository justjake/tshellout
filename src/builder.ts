import { Readable, Duplex } from 'node:stream'
import { Buffer } from 'node:buffer'
import { PathLike } from 'node:fs'

import * as errors from './errors'
import * as ast from './ast'
import evalCommand from './eval'
import { quote, toShellScript } from './toShellScript'

const readStream = async (stream: Readable): Promise<Buffer> => {
  const chunks = []

  for await (const chunk of stream) {
    chunks.push(Buffer.from(chunk))
  }

  return Buffer.concat(chunks)
}

export class CommandBuilder {
  private node: ast.Command

  constructor(node: ast.Command) {
    this.node = node
  }

  pipe(rhs: CommandBuilder) {
    return new CommandBuilder({ type: 'pipe', lhs: this.node, rhs: rhs.node })
  }

  and(rhs: CommandBuilder) {
    return new CommandBuilder({ type: 'and', lhs: this.node, rhs: rhs.node })
  }

  or(rhs: CommandBuilder) {
    return new CommandBuilder({ type: 'or', lhs: this.node, rhs: rhs.node })
  }

  redirectStdout(path: PathLike) {
    return new CommandBuilder({ type: 'redirectStdout', cmd: this.node, path })
  }

  redirectStderr(path: PathLike) {
    return new CommandBuilder({ type: 'redirectStderr', cmd: this.node, path })
  }

  redirectStdin(path: PathLike) {
    return new CommandBuilder({ type: 'redirectStdin', cmd: this.node, path })
  }

  writeStdin(data: Buffer) {
    return new CommandBuilder({ type: 'writeStdin', cmd: this.node, data })
  }

  async run() {
    const stdio = await evalCommand(this.node)
    stdio.stdin.end()

    const stdout = await readStream(stdio.stdout)
    const stderr = await readStream(stdio.stderr)

    let exitCode = null

    try {
      await stdio.terminated
      exitCode = 0
    } catch (err) {
      if (err instanceof errors.NonZeroExitCode) {
        exitCode = err.code
      } else {
        console.error(err)
      }
    } finally {
      if (stdio.stdout instanceof Duplex) {
        stdio.stdout.end()
      }

      if (stdio.stderr instanceof Duplex) {
        stdio.stderr.end()
      }
    }

    return {
      exitCode,
      stdout,
      stderr,
    }
  }

  toShellScript(): string {
    return toShellScript(this.node)
  }
}

type CommandTemplateArg =
  /** String arguments will be shell-escaped. */
  | string
  /** Command arguments will be interpolated literally. */
  | CommandBuilder
  /** Null and undefined arguments are omitted entirely. */
  | null
  | undefined

function unexpectedArgToString(arg: unknown): string {
  const constructorName = (
    arg as undefined | { constructor?: { name?: string } }
  )?.constructor?.name
  return `${typeof arg}/${constructorName || 'no constructor'}: ${String(arg)}`
}

function commandFromTaggedTemplateLiteral(
  template: string[],
  ...args: CommandTemplateArg[]
) {
  const executable = template
    .map((str, i) => {
      const arg = args[i]
      if (arg === null || arg === undefined) return str
      if (typeof arg === 'string') return `${str}${quote(arg)}`
      if (arg instanceof CommandBuilder) return `${str}${arg.toShellScript()}`
      else {
        arg satisfies never
        throw new TypeError(
          `Expected command template argument [${i}] to be string, Command, or nullable, but was ${unexpectedArgToString(
            arg
          )}`
        )
      }
    })
    .join(' ')

  return new CommandBuilder({ type: 'spawn', executable, args: [] })
}

function commandFromExecutableAndArgs(
  executable: string,
  ...args: string[]
): CommandBuilder {
  return new CommandBuilder({ type: 'spawn', executable, args })
}

export function command(executable: string, ...args: string[]): CommandBuilder
export function command(
  template: TemplateStringsArray,
  ...args: Array<CommandTemplateArg>
): CommandBuilder
export function command(
  executableOrTemplate: string | TemplateStringsArray,
  ...args: Array<string> | Array<CommandTemplateArg>
): CommandBuilder {
  if (typeof executableOrTemplate === 'string') {
    const stringArgs = args.map((arg, i) => {
      if (typeof arg !== 'string') {
        throw new TypeError(
          `Expected command argument [${i}] to be string, but was ${unexpectedArgToString(
            arg
          )}}`
        )
      }
      return arg
    })
    return commandFromExecutableAndArgs(executableOrTemplate, ...stringArgs)
  } else if (Array.isArray(executableOrTemplate)) {
    return commandFromTaggedTemplateLiteral(executableOrTemplate, ...args)
  } else {
    throw new TypeError(
      `Unexpected executable: ${unexpectedArgToString(executableOrTemplate)}`
    )
  }
}
