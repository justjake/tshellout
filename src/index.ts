import { command } from './builder'

export default command
