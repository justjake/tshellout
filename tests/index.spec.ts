import fs from 'node:fs/promises'
import path from 'node:path'
import os from 'node:os'

import command from '../src/index'
import { CommandBuilder } from '../src/builder'

const snapshotCommand = (cmd: CommandBuilder) => {
  expect(cmd.toShellScript()).toMatchSnapshot()
}

const testCommand = async (
  cmd: CommandBuilder,
  block: (
    res: Awaited<ReturnType<CommandBuilder['run']>>
  ) => void | Promise<void>
): Promise<void> => {
  const resFromEval = await cmd.run()
  await block(resFromEval)

  const shellScript = cmd.toShellScript()
  expect(shellScript).toMatchSnapshot()

  const resFromShell = await command(shellScript).run()
  try {
    await block(resFromShell)
  } catch (error) {
    console.error(
      'Bad shell script:\n',
      shellScript,
      '\nexitCode:',
      resFromShell.exitCode,
      '\nstdout:',
      resFromShell.stdout.toString(),
      '\nstderr:',
      resFromShell.stderr.toString(),
      '\n',
      String(error)
    )
    throw error
  }
}

describe('command runner', () => {
  it('should return stdout', async () => {
    const cmd = command('echo', 'hello world')

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(0)
      expect(res.stdout.toString().trim()).toEqual('hello world')
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })

  it('should return stderr', async () => {
    const cmd = command('>&2 echo', 'hello world')

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(0)
      expect(res.stdout.toString().trim()).toEqual('')
      expect(res.stderr.toString().trim()).toEqual('hello world')
    })
  })

  it('should return exit code', async () => {
    const cmd = command('exit', '42')

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(42)
      expect(res.stdout.toString().trim()).toEqual('')
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })

  it('should pipe stdout to stdin', async () => {
    const s = 'hello world'

    const cmd = command('echo', s)
      .pipe(command('tr', '-d', '\\r'))
      .pipe(command('tr', '-d', '\\n'))
      .pipe(command('wc', '-c'))

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(0)
      expect(res.stdout.toString().trim()).toEqual(`${s.length}`)
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })

  it('should chain commands with &&', async () => {
    const cmd = command('printf', 'hello ').and(command('printf', 'world'))

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(0)
      expect(res.stdout.toString().trim()).toEqual('hello world')
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })

  it('should short-circuit on failure with &&', async () => {
    const cmd = command('exit', '1').and(command('echo', 'hello world'))

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(1)
      expect(res.stdout.toString().trim()).toEqual('')
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })

  it('should chain commands with ||', async () => {
    const cmd = command('exit', '1').or(command('echo', 'hello world'))

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(0)
      expect(res.stdout.toString().trim()).toEqual('hello world')
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })

  it('should short-circuit on success with ||', async () => {
    const cmd = command('echo', 'hello world').or(command('exit', '1'))

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(0)
      expect(res.stdout.toString().trim()).toEqual('hello world')
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })

  it('should redirect stdout to a file', async () => {
    const folder = await fs.mkdtemp(
      path.join(os.tmpdir(), 'tshellout-test-stdout-')
    )
    const filepath = path.join(folder, 'stdout.txt')
    const cmd = command('echo', 'hello world').redirectStdout(filepath)

    try {
      await testCommand(cmd, async (res) => {
        expect(res.exitCode).toEqual(0)
        expect(res.stdout.toString().trim()).toEqual('')
        expect(res.stderr.toString().trim()).toEqual('')

        const content = await fs.readFile(filepath)
        expect(content.toString().trim()).toEqual('hello world')
      })
    } finally {
      await fs.rm(folder, { recursive: true, force: true })
    }
  })

  it('should redirect stderr to a file', async () => {
    const folder = await fs.mkdtemp(
      path.join(os.tmpdir(), 'tshellout-test-stderr-')
    )
    const filepath = path.join(folder, 'stderr.txt')
    const cmd = command('>&2 echo', 'hello world').redirectStderr(filepath)

    try {
      await testCommand(cmd, async (res) => {
        expect(res.exitCode).toEqual(0)
        expect(res.stdout.toString().trim()).toEqual('')
        expect(res.stderr.toString().trim()).toEqual('')

        const content = await fs.readFile(filepath)
        expect(content.toString().trim()).toEqual('hello world')
      })
    } finally {
      await fs.rm(folder, { recursive: true, force: true })
    }
  })

  it('should redirect a file to stdin', async () => {
    const folder = await fs.mkdtemp(
      path.join(os.tmpdir(), 'tshellout-test-stdin-')
    )
    const filepath = path.join(folder, 'stdin.txt')

    try {
      await fs.writeFile(filepath, 'hello world\n')

      const cmd = command('cat').redirectStdin(filepath)
      await testCommand(cmd, (res) => {
        expect(res.exitCode).toEqual(0)
        expect(res.stdout.toString().trim()).toEqual('hello world')
        expect(res.stderr.toString().trim()).toEqual('')
      })
    } finally {
      await fs.rm(folder, { recursive: true, force: true })
    }
  })

  it('should write to stdin', async () => {
    const cmd = command('cat').writeStdin(Buffer.from('hello world\n'))

    await testCommand(cmd, (res) => {
      expect(res.exitCode).toEqual(0)
      expect(res.stdout.toString().trim()).toEqual('hello world')
      expect(res.stderr.toString().trim()).toEqual('')
    })
  })
})

describe('command template literal', () => {
  it('shell-escapes interpolated strings', () => {
    snapshotCommand(command`echo ${'hello world'}`)
  })

  it('inserts interpolated commands literally', () => {
    const innerCmd = command`exit 0`
    const cmd = command`echo Running command: ${innerCmd}`
    snapshotCommand(cmd)
  })

  it('omits interpolated null | undefined', () => {
    snapshotCommand(command`echo ${null} hi`)
    snapshotCommand(command`echo ${undefined} bye`)
  })
})
