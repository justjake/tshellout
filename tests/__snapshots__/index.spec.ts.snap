// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command runner should chain commands with && 1`] = `"(printf 'hello ') && printf world"`;

exports[`command runner should chain commands with || 1`] = `"(exit 1) || echo 'hello world'"`;

exports[`command runner should pipe stdout to stdin 1`] = `"echo 'hello world' | tr -d '\\r' | tr -d '\\n' | wc -c"`;

exports[`command runner should redirect a file to stdin 1`] = `"(cat) < /var/folders/6f/y7zxg6h56r38gv53tqxxcwv00000gq/T/tshellout-test-stdin-v210dP/stdin.txt"`;

exports[`command runner should redirect stderr to a file 1`] = `"(>&2 echo 'hello world') 2> /var/folders/6f/y7zxg6h56r38gv53tqxxcwv00000gq/T/tshellout-test-stderr-appOr5/stderr.txt"`;

exports[`command runner should redirect stdout to a file 1`] = `"(echo 'hello world') > /var/folders/6f/y7zxg6h56r38gv53tqxxcwv00000gq/T/tshellout-test-stdout-GzEXzC/stdout.txt"`;

exports[`command runner should return exit code 1`] = `"exit 42"`;

exports[`command runner should return stderr 1`] = `">&2 echo 'hello world'"`;

exports[`command runner should return stdout 1`] = `"echo 'hello world'"`;

exports[`command runner should short-circuit on failure with && 1`] = `"(exit 1) && echo 'hello world'"`;

exports[`command runner should short-circuit on success with || 1`] = `"(echo 'hello world') || exit 1"`;

exports[`command runner should write to stdin 1`] = `"cat <<< "$(base64 -d <<< aGVsbG8gd29ybGQK)""`;

exports[`command template literal inserts interpolated commands literally 1`] = `"echo Running command: exit 0 "`;

exports[`command template literal omits interpolated null | undefined 1`] = `"echo   hi"`;

exports[`command template literal omits interpolated null | undefined 2`] = `"echo   bye"`;

exports[`command template literal shell-escapes interpolated strings 1`] = `"echo 'hello world' "`;
